File: 프로그래머스/프로그래머스_Lv3/야근 지수/야근지수.java
import java.util.PriorityQueue;
import java.util.Arrays;
import java.util.Collections;

class Solution {
    public long solution(int n, int[] works) {
        PriorityQueue<Integer> pq = new PriorityQueue<>(Collections.reverseOrder());
        Arrays.stream(works).forEach(each -> pq.add(each));
        for(int index = 0; index < n; index++) {
            int value = pq.poll();
            if(value == 0) return 0;
            pq.add(value-1);
        }

        long answer = 0;
        while(!pq.isEmpty()) {
            answer += Math.pow(pq.poll(), 2);
        }
        return answer;
    }
}

File: 프로그래머스/프로그래머스_Lv3/야근 지수/야근지수.js
function solution(n, works) {
  works.sort((a, b) => {
    return b - a;
  });

  for (let index = 0; index < n; index++) {
    let work = works.shift();
    if (work === 0) return 0;

    works.unshift(work - 1);
    works.sort((a, b) => b - a);
  }
  return works
    .map((each) => each * each)
    .reduce((acc, current) => acc + current, 0);
}
